cli/js/tests/unit_test_runner.ts

```diff
@@ -2,42 +2,180 @@
   result: number;
 }
 
-function permsToCliFlags(perms: Permissions): string[] {
-  return Object.keys(perms)
-    .map(key => {
-      if (!perms[key as keyof Permissions]) return "";
+/**
+ * Take a list of permissions and revoke missing permissions.
+ */
+async function dropWorkerPermissions(
+  requiredPermissions: Deno.PermissionName[]
+): Promise<void> {
+  const permsToDrop: Deno.PermissionName[] = [
+    "read",
+    "write",
+    "net",
+    "env",
+    "run",
+    "plugin",
+    "hrtime"
+  ];
+
+  for (const flag of requiredPermissions) {
+    if (!flag.length) continue;
+    const index = permsToDrop.indexOf(flag);
+    permsToDrop.splice(index, 1);
+  }
+
+  console.log("worker runner dropping permissions: ", permsToDrop);
+  for (const perm of permsToDrop) {
+    await Deno.permissions.revoke({ name: perm });
+  }
+}
 
-      const cliFlag = key.replace(
-        /\.?([A-Z])/g,
-        (x, y): string => `-${y.toLowerCase()}`
-      );
-      return `--allow-${cliFlag}`;
+async function workerRunnerMain(
+  addr: { hostname: string; port: number },
+  permissions: Deno.PermissionName[]
+): Promise<void> {
+  // Create reporter, then drop permissions to requested set
+  const conn = await Deno.connect(addr);
+  console.log("Test worker online", addr, permissions);
+  const socketReporter = new SocketReporter(conn);
+  await dropWorkerPermissions(permissions);
+
+  // Register unit tests that match process permissions
+  await registerUnitTests();
+
+  // Permissions dropped we're ready to execute tests
+  const results = await Deno.runTests({
+    failFast: false,
+    exitOnFail: false,
+    reporter: socketReporter
+  });
+  console.log("worker finished running tests", results.stats);
+  socketReporter.close();
+}
+
+function spawnWorkerRunner(addr: string, perms: Permissions): Deno.Process {
+  // run subsequent tests using same deno executable
+  const permStr = Object.keys(perms)
+    .filter((permName): boolean => {
+      return perms[permName as Deno.PermissionName] === true;
     })
-    .filter((e): boolean => e.length > 0);
+    .join(",");
+
+  const args = [
+    Deno.execPath(),
+    "run",
+    "-A",
+    "cli/js/tests/unit_test_runner.ts",
+    "--",
+    "--worker",
+    `--addr=${addr}`,
+    `--perms=${permStr}`
+  ];
+
+  const p = Deno.run({
+    args
+  });
+
+  return p;
 }
 
-function fmtPerms(perms: Permissions): string {
-  let fmt = permsToCliFlags(perms).join(" ");
+async function runTestsForPermissionSet(
+  reporter: Deno.ConsoleReporter,
+  perms: Permissions
+): Promise<PermissionSetTestResult> {
+  const permsFmt = fmtPerms(perms);
+  console.log(`Running tests for: ${permsFmt}`);
+  const addr = { hostname: "127.0.0.1", port: 4510 };
+  const addrStr = `${addr.hostname}:${addr.port}`;
+  const workerListener = Deno.listen(addr);
+
+  const workerProcess = spawnWorkerRunner(addrStr, perms);
+
+  // Wait for worker subprocess to go online
+  const conn = await workerListener.accept();
```

Comment:
There should be a timeout/some check here that worker runner actually starts properly. Otherwise this will hang the process

Reply:
I'm inclined to await for a single line output from stdout before accepting, that should fixed problem described above

Reply:
Doesn't work; if I close the `stdout` after receiving single message then tests fail; if I don't output from test runner is laying around in pipe